src/error_histogram_service.py

```diff
@@ -0,0 +1,183 @@
+        colors = [
+            'black',      # For values in the range < 0
+            'darkblue',   # Prussian Blue for values [0, 5]
+            'yellow',     # For values [5, 15]
+            'red',        # For values [15, 30]
+            'darkred'     # For values [30, 100]
+        ]
+        bounds = [-1, 0, 5, 15, 30, 100]
+
+        # Create the heatmap with its configurations
+        cmap = mc.ListedColormap(colors)
+        norm = mc.BoundaryNorm(bounds, cmap.N)
+        plt.figure(figsize=(data_matrix.shape[1] / 4, data_matrix.shape[0] / 4))
+        extent = (0, data_matrix.shape[1], 0, data_matrix.shape[0])
+        plt.imshow(data_matrix, cmap=cmap, norm=norm, aspect="auto", extent=extent)
+        plt.colorbar(label="Error Rate in %")
+        plt.subplots_adjust(left=0.2)
+
+        # Create horizontal lines and clinic labels for y axis
+        ticks = np.arange(len(data_matrix))
+        plt.hlines(ticks, xmin=0, xmax=data_matrix.shape[1], color='grey', linewidth=0.5)
+        label_ticks = ticks + 0.5
+        plt.yticks(ticks=label_ticks, labels=clinics[::-1], fontsize=8)
+        plt.xticks(ticks=np.arange(len(dates)), labels=dates, rotation=90, ha="left", fontsize=8)
+        plt.savefig('heatmap.png')
+
+    def _order_dict(self, data: dict) -> dict:
+        return dict(sorted(data.items(), key=lambda item: sum(item[1]), reverse=True))
+
+
+class ChartManager:
+    def __init__(self, mapper: ConfluenceNodeMapper, csv_paths: list = None,
+                 save_path: str = "error_rates_histogram.png", max_days: int = 42):
+        self.mapper = mapper
+        self.csv_paths = csv_paths if csv_paths is not None else []
+        self.save_path = save_path
+        self.max_days = max_days
+
+    def heat_map(self):
+        """
+        This method manages the collection of needed error rate data and initializes the Heatmap generation factory.
+        """
+        hm = HeatMapFactory()
+        data = {}
+
+        def process_path(path):
+            try:
+                dates, error_rates = Helper.read_error_rates(path)
+                error_rates = error_rates[-self.max_days:]
+                dates = dates[-self.max_days:]
+
+                clinic_id = Helper.get_clinic_num(path)
+                clinic_name = self.mapper.get_node_value_from_mapping_dict(clinic_id, "COMMON_NAME")
+                data[clinic_name] = error_rates
+                return data, dates
+            except Exception as e:
+                print(f"Error processing {path}: {e}")
+                return None
+
+        with ThreadPoolExecutor() as executor:
+            futures = {executor.submit(process_path, path): path for path in self.csv_paths}
+            for future in as_completed(futures):
+                result = future.result()
+                if result:
+                    data, dates = result
+
+        hm.plot(data, dates)
+        plt.savefig(self.save_path)
+
+
+class Helper:
```

Comment:
A helper class should contain common methods used by several classes. Here the helper is only used by ChartManager?

Reply:
I removed the Helper class and moved the functions into the ChartManager as protected methods